# Data Wrangling in Pandas
<img src='images/img_one.jpg' align=right>
The Journal of Data Science defines "data science" as almost everything that has something to do with data. In a job, this translates to using data to have an impact on the organization by adding value. Most commonly it is to use and apply the data to solve complex business problems. One of the most common steps taken in data science work is data wrangling. The following is a concise guide on how to go about exploring, manipulating and reshaping data in python using the <a href="https://pandas.pydata.org/pandas-docs/stable/">pandas</a> library.
 
We will explore a breast cancer data set (credits: <a href="https://archive.ics.uci.edu/ml/datasets/breast+cancer+wisconsin+(original)">UCI</a>) and use pandas to clean, reshape, massage and give us a clean data set, all of this will help dramatically increase the quality of our data. Note: Data quality is KEY for optimal performance with machine learning algorithms.

The following pandas functionalities will be covered:
1. Data exploration — columns, unique values in a column, describe, duplicates
2. Dealing with missing values — quantifying missing values per column, filling & dropping missing values
3. Reshaping data — one hot encoding, pivot tables, joins, grouping and aggregating
4. Filtering data
5. Other — Making descriptive columns, element-wise conditional operations



## Data Exploration
Let us begin by reading in our dataset (csv file) into pandas and displaying the column names along with their data types. Also take a moment to view the entire dataset.

In [2]:
import pandas as pd
pd.options.display.max_columns =None
pd.options.display.max_rows = 40
filename = 'data/breast_cancer_data.csv'
df = pd.read_csv(filename)

In [3]:
#start off by actually looking at your data set
df

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0       1000025              5.0                   1.0                      1   
1       1002945              5.0                   4.0                      4   
2       1015425              3.0                   1.0                      1   
3       1016277              6.0                   8.0                      8   
4       1017023              4.0                   1.0                      1   
..          ...              ...                   ...                    ...   
694      776715              3.0                   1.0                      1   
695      841769              2.0                   1.0                      1   
696      888820              5.0                  10.0                     10   
697      897471              4.0                   8.0                      6   
698      897471              4.0                   8.0                      8   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
0                    1                    2           1              3.0   
1                    5                    7          10              3.0   
2                    1                    2           2              3.0   
3                    1                    3           4              3.0   
4                    3                    2           1              3.0   
..                 ...                  ...         ...              ...   
694                  1                    3           2              1.0   
695                  1                    2           1              1.0   
696                  3                    7           3              8.0   
697                  4                    3           4             10.0   
698                  5                    4           5             10.0   

     normal_nucleoli  mitoses      class doctor_name  
0                1.0        1     benign     Dr. Doe  
1                2.0        1     benign   Dr. Smith  
2                1.0        1     benign     Dr. Lee  
3                7.0        1     benign   Dr. Smith  
4                1.0        1     benign    Dr. Wong  
..               ...      ...        ...         ...  
694              1.0        1     benign     Dr. Lee  
695              1.0        1     benign   Dr. Smith  
696             10.0        2  malignant     Dr. Lee  
697              6.0        1  malignant     Dr. Lee  
698              4.0        1  malignant    Dr. Wong  

[699 rows x 12 columns]

In [4]:
# What is the size of our dataset?
df.shape

(699, 12)

In [5]:
# Over here we see the columns names and their data types
df.dtypes

patient_id                 int64
clump_thickness          float64
cell_size_uniformity     float64
cell_shape_uniformity      int64
marginal_adhesion          int64
single_ep_cell_size        int64
bare_nuclei               object
bland_chromatin          float64
normal_nucleoli          float64
mitoses                    int64
class                     object
doctor_name               object
dtype: object

In the data we have the following columns as described by the source — Patient ID: id number, Clump Thickness: 1–10, Uniformity of Cell Size: 1–10, Uniformity of Cell Shape: 1–10, Marginal Adhesion: 1–10, Single Epithelial Cell Size: 1–10, Bare Nuclei: 1–10, Bland Chromatin: 1–10, Normal Nucleoli: 1–10, Mitoses: 1–10, Class: malignant or benign, Doctor name: 4 different doctors.

Based on this, we can assume that patient_id is a unique identifier, class is going to tell us whether the tumor is malignant (cancerous) or benign (not cancerous). The remaining columns are numeric medical descriptions of the tumor, except for the doctor_name which is a categorical feature.

Things to keep in mind — If our goal is to predict wether a tumor is cancerous or not based on the remaining features, we will have to one hot encode the categorical data and clean up the numerical data.
From our first output we see that bare_nuclei was read as an object data type although the description is numeric. Therefore we will need to change this.

To verify that our data matches up with the source we can use the describe option in pandas:

In [6]:
# This provides some statistics on the numerical data
df.describe()

patient_id  clump_thickness  cell_size_uniformity  \
count  6.990000e+02       698.000000            698.000000   
mean   1.071704e+06         4.416905              3.137536   
std    6.170957e+05         2.817673              3.052575   
min    6.163400e+04         1.000000              1.000000   
25%    8.706885e+05         2.000000              1.000000   
50%    1.171710e+06         4.000000              1.000000   
75%    1.238298e+06         6.000000              5.000000   
max    1.345435e+07        10.000000             10.000000   

       cell_shape_uniformity  marginal_adhesion  single_ep_cell_size  \
count             699.000000         699.000000           699.000000   
mean                3.207439           2.793991             3.216023   
std                 2.971913           2.843163             2.214300   
min                 1.000000           1.000000             1.000000   
25%                 1.000000           1.000000             2.000000   
50%                 1.000000           1.000000             2.000000   
75%                 5.000000           3.500000             4.000000   
max                10.000000          10.000000            10.000000   

       bland_chromatin  normal_nucleoli     mitoses  
count       695.000000       698.000000  699.000000  
mean          3.447482         2.868195    1.589413  
std           2.441191         3.055647    1.715078  
min           1.000000         1.000000    1.000000  
25%           2.000000         1.000000    1.000000  
50%           3.000000         1.000000    1.000000  
75%           5.000000         4.000000    1.000000  
max          10.000000        10.000000   10.000000

This neatly summarizes some statistical data for all numerical columns. For categorical data we can hand this by grouping together values:

In [7]:
# This aggreates the data by its column names, then we pass the aggregation function (size = count)
df.groupby(by =['class', 'doctor_name']).size()

class      doctor_name
benign     Dr. Doe        127
           Dr. Lee        121
           Dr. Smith      102
           Dr. Wong       108
malignant  Dr. Doe         58
           Dr. Lee         60
           Dr. Smith       74
           Dr. Wong        49
dtype: int64

## Dealing with missing values

With every dataset it is vital to evaluate the missing values. How many are there? Is it an error? Are there too many missing values? Does a missing value have a meaning relative to its context?
We can sum up the total missing values using the following:

In [8]:
#isna-> is not a number (shows if true or false)

df.isna()

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0         False            False                 False                  False   
1         False            False                 False                  False   
2         False            False                 False                  False   
3         False            False                 False                  False   
4         False            False                 False                  False   
..          ...              ...                   ...                    ...   
694       False            False                 False                  False   
695       False            False                 False                  False   
696       False            False                 False                  False   
697       False            False                 False                  False   
698       False            False                 False                  False   

     marginal_adhesion  single_ep_cell_size  bare_nuclei  bland_chromatin  \
0                False                False        False            False   
1                False                False        False            False   
2                False                False        False            False   
3                False                False        False            False   
4                False                False        False            False   
..                 ...                  ...          ...              ...   
694              False                False        False            False   
695              False                False        False            False   
696              False                False        False            False   
697              False                False        False            False   
698              False                False        False            False   

     normal_nucleoli  mitoses  class  doctor_name  
0              False    False  False        False  
1              False    False  False        False  
2              False    False  False        False  
3              False    False  False        False  
4              False    False  False        False  
..               ...      ...    ...          ...  
694            False    False  False        False  
695            False    False  False        False  
696            False    False  False        False  
697            False    False  False        False  
698            False    False  False        False  

[699 rows x 12 columns]

In [9]:
#Dealing with missing values? How many np.nan per column?

df.isna().sum() 

patient_id               0
clump_thickness          1
cell_size_uniformity     1
cell_shape_uniformity    0
marginal_adhesion        0
single_ep_cell_size      0
bare_nuclei              2
bland_chromatin          4
normal_nucleoli          1
mitoses                  0
class                    0
doctor_name              0
dtype: int64

Now that we have identified our missing values, we have a few options. We can fill them in with a certain value (zero, mean/max/median by column, string) or drop them by row. Since there are few missing values, we can drop the rows to avoid skewing the data in further analysis.

In [11]:
# # fill with zero
# df = df.fillna(0) -> isna all replaced by 0 (not recommended)

In [10]:
df = df.dropna(axis = 0, how = 'any')  #drop rows with any column having np.nan values

#Rename columns
#df.rename(index =str, columns = {'patient_id':'patient_id'})
df

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0       1000025              5.0                   1.0                      1   
1       1002945              5.0                   4.0                      4   
2       1015425              3.0                   1.0                      1   
3       1016277              6.0                   8.0                      8   
4       1017023              4.0                   1.0                      1   
..          ...              ...                   ...                    ...   
694      776715              3.0                   1.0                      1   
695      841769              2.0                   1.0                      1   
696      888820              5.0                  10.0                     10   
697      897471              4.0                   8.0                      6   
698      897471              4.0                   8.0                      8   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
0                    1                    2           1              3.0   
1                    5                    7          10              3.0   
2                    1                    2           2              3.0   
3                    1                    3           4              3.0   
4                    3                    2           1              3.0   
..                 ...                  ...         ...              ...   
694                  1                    3           2              1.0   
695                  1                    2           1              1.0   
696                  3                    7           3              8.0   
697                  4                    3           4             10.0   
698                  5                    4           5             10.0   

     normal_nucleoli  mitoses      class doctor_name  
0                1.0        1     benign     Dr. Doe  
1                2.0        1     benign   Dr. Smith  
2                1.0        1     benign     Dr. Lee  
3                7.0        1     benign   Dr. Smith  
4                1.0        1     benign    Dr. Wong  
..               ...      ...        ...         ...  
694              1.0        1     benign     Dr. Lee  
695              1.0        1     benign   Dr. Smith  
696             10.0        2  malignant     Dr. Lee  
697              6.0        1  malignant     Dr. Lee  
698              4.0        1  malignant    Dr. Wong  

[690 rows x 12 columns]

This allows us to drop rows with any missing values in them.

## Inspecting duplicates
To view repeating rows we can start off by looking at the number of unique values in each column.

In [12]:
# Here we list all columns
df.columns

Index(['patient_id', 'clump_thickness', 'cell_size_uniformity',
       'cell_shape_uniformity', 'marginal_adhesion', 'single_ep_cell_size',
       'bare_nuclei', 'bland_chromatin', 'normal_nucleoli', 'mitoses', 'class',
       'doctor_name'],
      dtype='object')

In [13]:
# the number changed cuz we dropped all the isna's
len(df)

690

In [14]:
# Its good to inspect your unique key identifier
df.nunique()

patient_id               637
clump_thickness           10
cell_size_uniformity      10
cell_shape_uniformity     10
marginal_adhesion         10
single_ep_cell_size       10
bare_nuclei               11
bland_chromatin           10
normal_nucleoli           10
mitoses                    9
class                      2
doctor_name                4
dtype: int64

We see here that although there are 690 rows, there are only 637 unique patient_id’s. This could mean that some patient appear more than once in the dataset. To isolate these patients and view their data, we use the following:

In [15]:
# This shows rows that show up more than once and have the exact same column values. 
#We are keeping the last value for each patient
df[df.duplicated(keep = 'last')]

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
168     1198641              3.0                   1.0                      1   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
168                  1                    2           1              3.0   

     normal_nucleoli  mitoses   class doctor_name  
168              1.0        1  benign     Dr. Lee

In [ ]:
# # This shows all instances where patient_id shows up more than once, but may have varying column values
df[df.duplicated(subset = 'patient_id', keep =False)].sort_values('patient_id')

In [ ]:
#Now that I have seen that there are some duplicates, I am going to go ahead and remove any duplicate rows

#df = df.drop_duplicates(subset = None, keep ='first')

In [ ]:
repeat_patients = df.groupby(by = 'patient_id').size().sort_values(ascending =False)
repeat_patients

This shows that one patient shows up in the data 6 times!

## Filtering data
If we want to remove patients that show up more that 2 times in the data set.

In [ ]:
filtered_patients = repeat_patients[repeat_patients > 2].to_frame().reset_index()
filtered_df = df[~df.patient_id.isin(filtered_patients.patient_id)]
filtered_df

If we did not have the tilde (“~”) we would get all individuals that repeat more than twice. By adding a tilde the pandas boolean series is reversed and thus the resulting data frame is of those that do NOT repeat more than twice.

## Reshaping data
The dataset has elements of categorical data in the “doctor_name” column. To feed this data into a machine learning pipeline, we will need to convert it into a one hot encoded column. This can be done with a sci-kit learn package, however we will do it in pandas to demonstrate the pivoting and merging functionality. Start off by creating a new dataframe with the categorical data.

In [ ]:
categorical_df = df[['patient_id','doctor_name']]
categorical_df

In [ ]:
# This specifies all rows (':') and column name 'doctor_count'
categorical_df.loc[:,'doctor_count'] = 1

In [ ]:
categorical_df

We add a column an extra column to identify which doctor a patient deals with. Pivot this table so that we only have numerical values in the cells and the columns become the doctors’ name. Then fill in the empty cells with 0.

In [ ]:
# hot-encoding is assigning numbers to categorical or subjective data (strings) for stats.
doctors_one_hot_encoded = pd.pivot_table( categorical_df,
                                  index = categorical_df.index, 
                                  columns = ['doctor_name'], 
                                   values = ['doctor_count'] )
doctors_one_hot_encoded = doctors_one_hot_encoded.fillna(0)
doctors_one_hot_encoded

Then drop the multiIndex columns:


In [ ]:
doctors_one_hot_encoded.columns = doctors_one_hot_encoded.columns.droplevel()
doctors_one_hot_encoded

We can now join this back to our main table. Typically a left join in pandas looks like this:

`leftJoin_df = pd.merge(df1, df2, on ='col_name', how='left')`

However we are joining on the index so we pass the “left_index” and “right_index” option to specify that the join key is the index of both tables



In [ ]:
combined_df = pd.merge(df, doctors_one_hot_encoded, left_index = True,right_index =True, how ='left')
combined_df

We can drop the column that we no longer need by the following:

In [ ]:
combined_df = combined_df.drop(columns=['doctor_name'])
combined_df

## Row-wise Operations
Another key component in data wrangling is having the ability to conduct row-wise or column wise operations. Examples of this are; rename elements within a column based on its value and create a new column that yields a specific value based on multiple attributes within the row.

For this example lets create a new column that categorizes a patients cell as normal or abnormal based on its attributes. We first define our function and the operation that it will be doing.

In [ ]:
def celltypelabel(x):
    if ((x['cell_size_uniformity'] < 5) &      (x['cell_shape_uniformity'] < 5)):
        
        return('normal')
    else:
        return('abnormal')

Then we use the pandas apply function to run the celltypelabel(x) function on the dataframe.

In [ ]:
combined_df['cell_type_label'] = combined_df.apply(lambda x: celltypelabel(x), axis=1)

In [ ]:
combined_df

## Conclusion
Although some of these data manipulation steps can be done in SAS and excel. Doing it in python not only allows you to connect the data to vast open source resources in computer vision, machine and deep learning, but also for ETL automation purposes and more.

# Homework
1. Hot encode a new column in this dataset for cancerous (1) or not cancerous (0).